rbd/snapshot_namespace.go

```diff
@@ -0,0 +1,76 @@
+// +build !luminous
+//
+// Ceph Mimic introduced rbd_snap_get_namespace_type().
+
+package rbd
+
+// #cgo LDFLAGS: -lrbd
+// #include <rbd/librbd.h>
+import "C"
+import (
+	"unsafe"
+
+	"github.com/ceph/go-ceph/internal/retry"
+)
+
+// SnapNamespaceType indicates the namespace to which the snapshot belongs to.
+type SnapNamespaceType C.rbd_snap_namespace_type_t
+
+const (
+	// SnapNamespaceTypeUser indicates that the snapshot belongs to user namespace.
+	SnapNamespaceTypeUser = SnapNamespaceType(C.RBD_SNAP_NAMESPACE_TYPE_USER)
+
+	// SnapNamespaceTypeGroup indicates that the snapshot belongs to group namespace.
+	// Such snapshots will have associated group information.
+	SnapNamespaceTypeGroup = SnapNamespaceType(C.RBD_SNAP_NAMESPACE_TYPE_GROUP)
+
+	// SnapNamespaceTypeTrash indicates that the snapshot belongs to trash namespace.
+	SnapNamespaceTypeTrash = SnapNamespaceType(C.RBD_SNAP_NAMESPACE_TYPE_TRASH)
+)
+
+// GetSnapNamespaceType gets the type of namespace to which the snapshot belongs to,
+// returns error on failure.
+// Implements:
+//  int rbd_snap_get_namespace_type(rbd_image_t image, uint64_t snap_id, rbd_snap_namespace_type_t *namespace_type)
+func (image *Image) GetSnapNamespaceType(snapID uint64, namespace *SnapNamespaceType) error {
+	if err := image.validate(imageIsOpen); err != nil {
```

Comment:
@agarwal-mudit isn't possible to return the list of snapshotinfo for an image? And also how to get the snapid to call this function?

Reply:
Can you also provide the steps to get all the snapshot info (which are not in trash and which are in trash)

Reply:
This is not a command implementation so user needs to follow these steps:

1. Call GetSnapshotNames(), this will return **snapInfo** for all the snapshots (including the one in Trash)
2. Using GetSnapNamespaceType() iterated all the snashot IDs you will get (**snapInfo.Id**) and check which snapshots belong to RBD_SNAP_NAMESPACE_TYPE_TRASH.
3. Only for those snapshots you need to call GetSnapNamespaceTrash() to verify the snapshot names.

For more details, either check do_list_snaps() [here](https://github.com/ceph/ceph/blob/584baf132fb077e8e0c84cfc3edd1615bd574698/src/tools/rbd/action/Snap.cc#L25), or the test case I have added for this PR.

Reply:
> This is not a command implementation so user needs to follow these steps:
> 
> 1. Call GetSnapshotNames(), this will return **snapInfo** for all the snapshots (including the one in Trash)
> 2. Using GetSnapNamespaceType() iterated all the snashot IDs you will get (**snapInfo.Id**) and check which snapshots belong to RBD_SNAP_NAMESPACE_TYPE_TRASH.
> 3. Only for those snapshots you need to call GetSnapNamespaceTrash() to verify the snapshot names.
> 
> For more details, either check do_list_snaps() [here](https://github.com/ceph/ceph/blob/584baf132fb077e8e0c84cfc3edd1615bd574698/src/tools/rbd/action/Snap.cc#L25), or the test case I have added for this PR.

Isn't this heavy? if I have 500 snapshots in the trash for an rbd image, to get original names, I need to make 500*2=1000 calls right? is it possible to provide a single api that returns the []struct as CLI does?

Reply:
Even if we implement such API, it will do the same internally.

@phlogistonjohn / @nixpanic comments please?

Reply:
I just double checked what do_list_snaps() does and it is pretty much how @agarwal-mudit describes it. So it's no more heavyweight than what the cli was doing (and the cli was emitting text which then needed to get parsed). 
I am not strictly opposed to adding some convenience apis to the library, but they need to be obvious or be widely needed. I don't see either of those applying in this case right now. So I'm disinclined to add a function like that.